scripts/beesd.in

```diff
@@ -91,7 +91,7 @@ MNT_DIR="${MNT_DIR:-$WORK_DIR/mnt/$UUID}"
 BEESHOME="${BEESHOME:-$MNT_DIR/.beeshome}"
 BEESSTATUS="${BEESSTATUS:-$WORK_DIR/$UUID.status}"
 DB_SIZE="${DB_SIZE:-$((8192*AL128K))}"
-MOUNT_OPTIONS="${MOUNT_OPTIONS-}"
+MOUNT_OPTIONS="${MOUNT_OPTIONS:-}"
```

Comment:
What is achieved here?  It looks like MOUNT_OPTIONS will be the empty string if it is not defined, whether or not the : is present.

Is there a `set -u` somewhere?

Reply:
If there's no `set -u` then there's no need for this line at all, except as a sort of documentation that `MOUNT_OPTIONS` is a variable that can be set that happens to have no default value.

Reply:
There no `set -u`, so yeah, only documentation for `MOUNT_OPTIONS` is required.
There's documentation in `beesd.conf.sample`, though I'm not sure whether I should also add one in `beesd.in`.

Reply:
I'm not sure if compression options are even a per mount point option, I think it's used globally for the device, where ever that happens to be mounted.

Reply:
Then it’s probably just me misunderstanding how compression in Btrfs works.

Reply:
I think compression mode is a global setting per file system, it cannot be set individually per mount point or per sub volume. But you can force compression on per subvolume using `chattr` on the subvol directory. It will be inherited by new files and folders then. But it's not possible to force compression off that way. So it's either forced on/off globally per fs, automatic via heuristics, or forced on per subvolume/file/directory.

I think what `man 5 btrfs` means with
> You can have a single btrfs mount point that has some files that are uncompressed, some that are compressed with LZO, some with ZLIB, for instance (though you may not want it that way, it is supported)
does not mean that it is a per subvol setting but just that you can mix and match between the settings with each mount and it will happily accept having different compression modes persisted to different files at the same time in the FS.

Also refer to "Can I mount subvolumes with different mount options?" here: https://btrfs.wiki.kernel.org/index.php/FAQ
> The generic mount options can be different for each subvolume, see the list below. Btrfs-specific mount options cannot be specified per-subvolume, but this will be possible in the future (a work in progress).

Reply:
> Seems that I didn't add that in `beesd.conf.sample`

Actually, you removed it in one commit.

Reply:
So, is it a good idea to keep it?

Perhaps user might want to use [mount options](https://btrfs.wiki.kernel.org/index.php/Manpage/btrfs(5)#MOUNT_OPTIONS) like discard=async or max_inline=bytes.

And yeah, I think the user could also employ noatime and lazytime, perhaps also nodev, noexec and nosuid if they are being pedantic.

Also, I wonder if what’s the effect of using compress and compress-force?

If both is specified with the same algorithm and level, will btrfs respect compress-force and enforces it upon all mount points?

Reply:
> Perhaps user might want to use [mount options](https://btrfs.wiki.kernel.org/index.php/Manpage/btrfs(5)#MOUNT_OPTIONS) like discard=async, max_inline=bytes or flushcommit.

These are all btrfs specific and thus cannot be set per mount point - and the general use case is that the btrfs is already mounted somewhere else by the system - so it's futile to set these options here, or may have unexpected side effects because it secretly changed the options for your other active mounts of the FS, too.

> And yeah, I think the user could also employ noatime and lazytime, perhaps also nodev, noexec and nosuid if they are being pedantic.

OTOH, these are documented to work per subvolume/mountpoint. It may be useful to be pedantic by default for the auto-determined mount.

> If both is specified with the same algorithm and level, will btrfs respect compress-force and enforces it upon all mount points?

That's how I'm currently reading the documentation. @Zygo may know more because he's much more closely involved in the kernel development of btrfs. Docs may just be out of date.

Reply:
> And yeah, I think the user could also employ noatime

The idea here is that you don't want bees update atime when it starts opening random files it picks up from reverse extent lookup. OTOH, I think bees already opens all files with noatime.

I'd simply just not mess with mount options at all and let it use whatever the user already specified in fstab. After all, bees just mounts a file system that is most likely already mounted somewhere else through fstab, not specifying any mount options on that secondary mount would go with whatever that entry provided - not overriding any options.

The bees script should probably just mount with `rw,noatime` - and that's it: It needs write access, and it has no business in updating anyones atimes. The systemd service could be locked down by providing options to restrict file access to certain directories and access to a limited set of syscalls - if that makes any sense. bees is really not a service that accepts external unvalidated user input - so it would probably be overkill. Maybe it could drop root privileges but keep `CAP_FOWNER` or whatever might be needed to open any file in the mount point and use the ioctls.

Reply:
Systemd supports Protect* options that can:
 - Disable access to home
 - Unshare /tmp and /var/tmp
 - Mount root filesystem read only
 - Forbidden kernel module loading or changing parameter
 
 etc.

Reply:
@kakra I think the following capability is also required by beesd:
 - `CAP_SYS_ADMIN` for mounting the btrfs
 - `CAP_DAC_READ_SEARCH` for bypassing file read permission checks and directory read and execute permission checks
 - `CAP_DAC_OVERRIDE` for bypassing file read, write, and execute permission checks.

`CAP_FOWNER` explicitly stated that it does not cover what `CAP_DAC_READ_SEARCH` and `CAP_DAC_OVERRIDE` cover, so I think these two are also required.

Reply:
@Zygo Is btrfs mount option global or only apply to the mount point it is used?

For example, if I specify `discard=async, max_inline=1000,flushcommit,compress=zstd:15`, would this affect all mountpoints which is backed by the same filesystem or just the mount point of which these options are specified?

Also, I wonder if what’s the effect of using both `compress` and `compress-force` on different mount points?

If both is specified with the same algorithm and level, will btrfs respect compress-force and enforces it upon all mount points?

Reply:
> * `CAP_SYS_ADMIN` for mounting the btrfs

The bees binary doesn't mount anything at all... It's only the wrapper script. That leaves us with the decision if the bees wrapper script itself should already have it's root permissions dropped...

I'm running bees without the wrapper script, btw. So I'd vote for the binary dropping its permissions by itself. But that's just an idea which needs a separate implementation. I was only thinking about what we could do from the systemd service already to prepare it better for future improvements and which options we have.

Reply:
Could even do the mounting in another service and join its mount namespace, but that is too complicated IMHO.

Dropping unnecessary capability in binary sounds like the right solution.

Reply:
I'd even vote for having the mounting and hash file management within the bees binary but AFAIR @Zygo doesn't want those kind of external libs as a dependency.

Reply:
@Zygo I’ve added default value for `MOUNT_OPTIONS` which I consider as reasonable.
Feel free to correct me if you think default value is off.

Reply:
> > If both is specified with the same algorithm and level, will btrfs respect compress-force and enforces it upon all mount points?
> 
> That's how I'm currently reading the documentation. @Zygo may know more because he's much more closely involved in the kernel development of btrfs. Docs may just be out of date.

In some cases (e.g. `subvol=` mounts) it's the first mount that specifies the option; in other cases (e.g. `mount -oremount`, it's the last.  Either way, in the kernel there is only one storage location for the `compress` option (including `force-` and compression level), so there will be only one setting for the filesystem in effect at any time. 

The only thing that can be changed per-file is the compression algorithm (i.e. zlib/zstd/lzo).  btrfs will store compression level in the xattr, but the kernel will not (yet) read it.

I'd say the expected behavior is undefined when two btrfs mount points a) both specify a `compress` option and b) specify _different values_ of that option.  Best practice is to use the option once for the entire kernel instance (i.e. don't set it in bees, set it when the filesystem is mounted somewhere else), or make sure it's exactly the same everywhere (which would require aligning `beesd` with whatever other mounts you have going on).

Similar restrictions exist for `discard=async`, `max_inline`, etc--there is only one value stored in the kernel for the entire filesystem, so mount commands that specify different options will change them for the entire filesystem.

So I wouldn't put any compress= option in bees by default.

`noatime,nosuid,nodev,noexec` are harmless, but have little to no effect on bees.  It would be better to lock down the `/run/bees` directory with `chmod 700` or similar to make sure it's not accessible to other users, or use a private mount namespace to run bees, so that bees has exclusive access.

Reply:
@Zygo Thank you very much for the clarification.

So with the private mount namespace enabled in systemd unit, it seems having MOUNT_OPTIONS in the configuration is unnecessary.

I will modify this PR to remove it.